theseus/optimizer/dense_linearization.py

```diff
@@ -72,3 +72,6 @@ def _atb_impl(self) -> torch.Tensor:
 
     def Av(self, v: torch.Tensor) -> torch.Tensor:
         return self.A.bmm(v.unsqueeze(2)).squeeze(2)
+
+    def diagonal_scaling(self, v: torch.Tensor) -> torch.Tensor:
+        return v * self._AtA.diagonal(dim1=1, dim2=2)
```

Comment:
How do we add `damping` to `self._AtA`? Is there another variable to cache the hessian matrix?

Reply:
The order shouldn't matter here because `self._AtA.diagonal` is also a vector, not a matrix (see attached example), so this is just elementwise product. 

![image](https://user-images.githubusercontent.com/4759586/205426157-0b03233f-db14-46de-8311-0d728cf3ba23.png)

We add the damping in the dense solver before solving, [here](https://github.com/facebookresearch/theseus/blob/main/theseus/optimizer/linear/dense_solver.py#L38). We don't currently cache (AtA + \lambda * diag(AtA)), although I don't think we are calling this more than once per iteration.

Reply:
Thanks!